### ISSUE TITLE
[Feature Request] Sorted Group By

### ISSUE THREAD

**Issue by jezumbro** [more-itertools/more-itertools#695]
### Description
Enhancing the `groupby` function for unsorted data. 

Using the initial group by function works however if the data is unsorted by the key you are filtering upon you will iterate over the group again. This creates some issues when creating a lookup or trying to group by a specific key.

### References
https://stackoverflow.com/questions/24666121/grouping-and-sorting-lists-using-python

### Examples
```
from itertools import groupby

iterable = (
  (0, 1),
  (0, -1),
  (1, 2),
  (1, 5),
  (0, 2)
)

expected = (
  (0, ((0, 1), (0, -1), (0, 2))), 
  (1, ((1, 2), (1, 5))), 
)

def by_first_item(x):
    return x[0] 

groupby_output = tuple(
    (k, tuple(v)) for k, v in groupby(iterable, key=by_first_item)
)

groupby_output = (
  (0, ((0, 1), (0, -1))), 
  (1, ((1, 2), (1, 5))), 
  (0, ((0, 2),))
)

sorted_group_by_output = tuple(
    (k, tuple(v)) for k, v in sorted_groupby(iterable, key=by_first_item)
)

sorted_group_by_output = (
  (0, ((0, 1), (0, -1), (0, 2))), 
  (1, ((1, 2), (1, 5))), 
)
```

**Comment by nanouasyn** [comment]
I have mixed feelings about this proposal. Indeed, very often we want to group not consecutive elements corresponding to the predicate, but all elements corresponding to the predicate. In this case, we are forced to call `sorted` and `groupby` with the same `key`, that is too verbose and does not allow us to use a single lambda function.

But let's assume that we wrote `sorted_groupby` somehow like this:

```python
def sorted_groupby(iterable, key=None):
    if key is None:
        key = lambda x: x
    yield from groupby(sorted(iterable, key=key), key=key)
```

When the first element is received from the output iterator, the input iterator is instantly consumed and its elements are stored in memory until we complete the iteration. And this sorting when getting the first element takes almost all the time. In other words, despite the fact that we return an iterator, iterating over it does not give us any benefits that iterators are designed for. However, we still have to pay: the resulting iterator can be iterated only once. And we have to process all the elements of each group in a row, which sometimes leads to unexpected errors. And all keys should be comarable.

Let's take a look at an alternative that doesn't use iterators:

```python
def dict_groupby(iterable, key=None):
    if key is None:
        key = lambda x: x
    d = {}
    for item in iterable:
        k = key(item)
        if k not in d:
            d[k] = [item]
        else:
            d[k].append(item)
    return d
```

It has a complexity of O(n), as opposed to O(n * log(n)) in the previous case. The elements from the input iterator are handled only once and immediately saved into the corresponding groups, and we get all the benefits of the dictionary including the ability to iterate through groups in any order any number of times. This covers most of the use cases of `sorted_groupby`, and works better. But it's not "itertool", which means it doesn't belong in this library.

By the way, if your proposal is still accepted, it would be good if the name follows the convention `unique_everseen`/`unique_justseen`, `duplicates_everseen`/`duplicates_justseen`, if it's possible.

**Comment by jezumbro** [comment]
@nanouasyn thanks for the feedback.

I do like the proposal of using a dictionary instead of using the `sorted` method. Especially when it comes with the performance enhancements of moving from O(n * log(n)) to O(n).

Here's the refactor I picture
```
def foobar(iterable, key):
    d = defaultdict(list)
    for item in iterable:
        d[key(item)].append(item)
    yield from d.items()
```

> This covers most of the use cases of sorted_groupby, and works better. But it's not "itertool", which means it doesn't belong in this library.

This method is allowing the `itertools.groupby` function to take unsorted data and actually group it together by key instead of creating multiple instances of the same key of unsorted data. (There is a test proving this point using the same data.) Since we are extending functionality around `itertools.groupby`, I do believe this method belongs in the library.

> By the way, if your proposal is still accepted, it would be good if the name follows the convention unique_everseen/unique_justseen, duplicates_everseen/duplicates_justseen, if it's possible.

Naming is hard... What do you think of `groupby_key`, `extended_groupby`? Sorted no longer makes sense, so I would be open to suggestions here.

**Comment by nanouasyn** [comment]
It is sad that in this case the returned iterator is deceptive. We keep all the elements in a dictionary in memory anyway. A user who wants to get a permanent storage of groups will be forced to create a new dictionary and move elements from our dictionary to his own. That is, to do additional useless work, just because we didn't give him our own dictionary. In addition, such an interface creates a deceptive impression that we are dealing with a kind of lazy calculation, although after receiving the first group, all real calculations are completed. It seems to me that we should take an example from the `sorted` function in standard library. `sorted`, unlike `filter` or `map`, does not return an iterator because it 
is impossible to sort lazily, and `sorted` does not want to give the impression that it is possible. So, it returns a list as is.

I don't know much English, so if the function was called `groupby_everseen`, I would understand that the function is related to `groupby` in the same way that `unique_everseen` is related to `unique_justseen`. Вut I don't know how bad it sounds :) To me, `groupby_key` sounds like a function that groups by key. But after all, `groupby` already groups by key. If I were to encounter `extended_groupby`, I would expect that this function is related to `groupby` in the same way that the `islice_extended` function is related to `islice`. As if `extended_groupby` can additionally handle cases that `groupby` is not able to handle. But this relation is completely different. So if not `groupby_everseen`, then I don't know :)

**Comment by jezumbro** [comment]
@bbayles yup didn't see that method before but I will be sure to use it in future work!

I'll remove the PR
